### DISCUSSION TITLE
Job hangs when object is put in map instead of string

### DISCUSSION THREAD

**Discussion by asherber** [quartznet/quartznet#1669] [Q&A]
**Describe the bug**

Given (1) a `JobStoreTX` initialized with `UseProperties = true` and (2) a job with `DisallowConcurrentExecution` and `PersistJobDataAfterExecution` which mistakenly stores an object in the `JobDataMap` instead of a string, the job will hang on its first run and will never recover.

**Version used**

3.4.0

**To Reproduce**

Using a scheduler with a `JobStoreTX` with `UseProperties=true`, schedule the following job to run every 5 seconds:

```csharp
[DisallowConcurrentExecution]
[PersistJobDataAfterExecution]
public class TestJob : IJob
{
    public Task Execute(IJobExecutionContext context)
    {
        var current = DateTime.Now;
        var last = context.MergedJobDataMap.GetDateTimeValue("LastRunTime");
        Console.WriteLine($"Job fired at {current}");
        Console.WriteLine($"Last run was {last}\n");
        context.JobDetail.JobDataMap.Put("LastRunTime", current);
        
        return Task.CompletedTask;
    }
}
```

**Expected behavior**

Since I've stored an object in a map that only expects string properties, there should be an exception thrown, which means that the job should run repeatedly with `last` always equal to `default(DateTime)`. Instead, the job runs exactly once and is never called again. 

**Additional context**

If `DisallowConcurrentExecution` is removed, behavior is as expected. Changing the schedule time of the job doesn't seem to have an effect.

**Comment by jafin** [comment]
Hi @asherber 

Here is what I'm noticing.

In my test, setup similar to what you posted,  the job is not hanging, it is throwing an exception, but it's not visible initially because of a retry delay loop. 

An exception is raised and will be passed to the ScheduleListener, so you will need a SchedulerListener registered to the Scheduler  to action this exception (I think).

ie 

```c#
public class SchedulerListener: ISchedulerListener {
    public Task SchedulerError(string msg, SchedulerException cause, CancellationToken cancellationToken = default)
        {
            Console.WriteLine($"{DateTime.Now}: Scheduler error {msg} {cause.Message}");
            return Task.CompletedTask;
        }
}
...
sched.ListenerManager.AddSchedulerListener(new SchedulerListener());

```
The defaults for the JobStoreTX set retries at 15 seconds, and a retry count of 4 before the Listener will be notified.
It will be 60 seconds before the first exception notification is pushed.


```mermaid
sequenceDiagram
    participant J as Job
    participant JSS as JobStoreSupport
    participant N as SchedulerListener
    loop Every retry
      J->>JSS: throw JobPersistenceException
     JSS->>JSS:if retry%RetryableActionErrorLogThreshold == 0
     alt isThreshold(15secx4)
        JSS->>N:NotifySchedulerListenersError
    end
    end
```

After running through this, slightly tangential, I wanted to reduce the retry interval for the test,  I think there is a bug in quartz as it is looking at the property, `quartz.scheduler.dbFailureRetryInterval` which was removed in 3.x,  but it appears to still read the value, but should be reading from `quartz.jobStore.dbRetryInterval`  to get the interval. I'll investigate a bit more and raise a separate issue.

**Comment by asherber** [comment]
Thanks @jafin, you're right about the error being picked up by the `SchedulerListener`. I admit that this is not the behavior I expect. For example, given the `TestJob` class above, I would expect the job to start successfully each time, output the console lines, and then throw an exception at the end (which the listener would eventually pick up). Instead, it seems that somehow Quartz just keeps retrying the `JobDataMap.Put()`, since I get a new notification from the listener every minute.

Is this the behavior as designed? And does it happen for any exception thrown during job execution?

**Comment by lahma** [comment]
Some of the behavior might be rooted in the fact that Java, and thus Quartz Java version (which Quartz.NET is based on) has different exception handling semantics due to language differences. In Java version `Execute` can be forced to have `JobExecutionException` as only allowed exception type to be thrown. 

This leads to the fact that any exception thrown is considered unhandled **unless** your job has try-catch and throws the `JobExecutionException` which can control retry etc. So Jave version can signal the handling via exceptions and .NET has to rely on end user doing the exception wrapping etc without compiler help.
